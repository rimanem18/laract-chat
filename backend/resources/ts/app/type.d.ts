export type PromiseState = 'idle' | 'loading' | 'rejected'

export type Message = {
  id: number
  group_id: number
  user_id: number
  name: string
  content: string
  created_at: string
  roles: string[]
}
export type Group = {
  id: number
  name: string
  roles: string[]
}
export type Role = {
  id: number
  name: string
  color: string
}
export type RoleUser = {
  user_id: number
  role_id: number
}
export type RoleGroup = {
  group_id: number
  role_id: number
}

export type Groups = {
  byId: Record<string, Group>
  allIds: string[]
}

export type Roles = {
  byId: Record<string, Role>
  allIds: string[]
}

export type Messages = {
  byId: Record<string, Message>
  allIds: string[]
}

/**
 * Payloads
 *
 * バックエンドから送られてくるデータの型
 */
export type GroupsPayload = {
  public_groups: {
    id: number
    name: string
  }[]
  private_groups: {
    id: number
    name: string
  }[]
  role_group: {
    role_id: number
    group_id: number
  }[]
}
export type RoleGroupPayload = {
  role_group: {
    group_id: number
    role_id: number
  }[]
}
export type RolesPayload = {
  id: number
  name: string
  color: string
}
export type MessagePayload = {
  id: number
  group_id: number
  content: string
  created_at: string
  user_id: number
  name: string
}
export type RoleUserPayload = {
  user_id: number
  role_id: number
}
