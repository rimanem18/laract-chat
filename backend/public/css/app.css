@import url(https://fonts.googleapis.com/css?family=Nunito);
/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */html{-webkit-text-size-adjust:100%;line-height:1.15}body{margin:0}main{display:block}h1{font-size:2em;margin:.67em 0}hr{box-sizing:content-box;height:0;overflow:visible}pre{font-family:monospace,monospace;font-size:1em}a{background-color:transparent}abbr[title]{border-bottom:none;text-decoration:underline;-webkit-text-decoration:underline dotted;text-decoration:underline dotted}b,strong{font-weight:bolder}code,kbd,samp{font-family:monospace,monospace;font-size:1em}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}img{border-style:none}button,input,optgroup,select,textarea{font-family:inherit;font-size:100%;line-height:1.15;margin:0}button,input{overflow:visible}button,select{text-transform:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]::-moz-focus-inner,[type=reset]::-moz-focus-inner,[type=submit]::-moz-focus-inner,button::-moz-focus-inner{border-style:none;padding:0}[type=button]:-moz-focusring,[type=reset]:-moz-focusring,[type=submit]:-moz-focusring,button:-moz-focusring{outline:1px dotted ButtonText}fieldset{padding:.35em .75em .625em}legend{box-sizing:border-box;color:inherit;display:table;max-width:100%;padding:0;white-space:normal}progress{vertical-align:baseline}textarea{overflow:auto}[type=checkbox],[type=radio]{box-sizing:border-box;padding:0}[type=number]::-webkit-inner-spin-button,[type=number]::-webkit-outer-spin-button{height:auto}[type=search]{-webkit-appearance:textfield;outline-offset:-2px}[type=search]::-webkit-search-decoration{-webkit-appearance:none}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}details{display:block}summary{display:list-item}[hidden],template{display:none}
.btn,.btn--red{border:1px solid #999;border-radius:5px;margin:.5em;padding:.5em}.btn--red{background:red;color:#f2f2f2}.container{margin:0 auto;padding:0 5px;width:100%}@media screen and (min-width:768px){.container{max-width:1100px;padding:0 15px}}.flex{display:block}@media screen and (min-width:768px){.flex{display:flex;flex-wrap:wrap}}.flex__item--1{flex:0 0 8.3333333333%;min-width:8.3333333333%}.flex__item--2{flex:0 0 16.6666666667%;min-width:16.6666666667%}.flex__item--3{flex:0 0 25%;min-width:25%}.flex__item--4{flex:0 0 33.3333333333%;min-width:33.3333333333%}.flex__item--5{flex:0 0 41.6666666667%;min-width:41.6666666667%}.flex__item--6{flex:0 0 50%;min-width:50%}.flex__item--7{flex:0 0 58.3333333333%;min-width:58.3333333333%}.flex__item--8{flex:0 0 66.6666666667%;min-width:66.6666666667%}.flex__item--9{flex:0 0 75%;min-width:75%}.flex__item--10{flex:0 0 83.3333333333%;min-width:83.3333333333%}.flex__item--11{flex:0 0 91.6666666667%;min-width:91.6666666667%}.group{-ms-overflow-style:none;height:65vh;margin:.2em;overflow:scroll;overflow-y:scroll;padding:0;scrollbar-width:none}.group::-webkit-scrollbar{display:none}.group__item{list-style:none;margin:0}.group .group__link,.group .group__link--is-active{border-radius:10px;display:inline-block;padding:.75em;text-decoration:none;width:100%}.group__link,.group__link:hover{color:#333}.group__link--is-active{background:#3fb455;color:#f2f2f2;font-weight:700}.group__link--is-active:hover{color:#f2f2f2}.icon-btn,.icon-btn--check,.icon-btn--close{background:transparent;border:0;cursor:pointer;margin:0;padding:0}.icon-btn{color:#333}.icon-btn--check{color:#38c172;margin-left:.2em;padding:1em}.icon-btn--close{color:#f2f2f2;padding:.5em;position:absolute;right:.5em;top:.5em}.input{margin:.5em;padding:.5em}.main{box-sizing:border-box;margin:15px}@media screen and (min-width:768px){.main{width:calc(100vw - 360px)}}.message{height:65vh;overflow:scroll}.message__item{border-top:1px solid #ccc;padding-top:.5em}.message__note{font-size:75%;text-align:center}.modal{display:block}.modal__title{background:#3fb455;color:#f2f2f2;margin:0;padding:1em;text-align:center}.modal__content{margin:0 auto;padding:.5em;width:90%}.modal__input{background:transparent;border:0;display:inline-block;margin:0;padding:.25em;text-align:center}.modal__input:active,.modal__input:focus,.modal__input:focus-visible{border:0;border-bottom:1px solid #ddd}.modal__footer{bottom:0;left:3%;position:absolute}.post-form{bottom:0;width:100%}.post-form__input{width:80%}@media screen and (min-width:768px){.post-form__input{width:90%}}.scroll-btn{bottom:20%;position:absolute;right:10%}.sidebar{background:#fff;box-sizing:border-box;padding:15px}@media screen and (min-width:768px){.sidebar{width:280px}}body,html{overflow:hidden}.loader{left:45%;position:absolute;top:45%}.loader-overlay{background:rgba(0,0,0,.7);height:100vh;left:0;position:absolute;top:0;width:100vw}
