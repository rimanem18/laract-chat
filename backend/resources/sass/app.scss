// Fonts
@import url('https://fonts.googleapis.com/css?family=Nunito');

@import '~normalize.css/normalize.css';

// Variables
@import 'variables';

// Bootstrap
// @import '~bootstrap/scss/bootstrap';
@import 'btn';
@import 'container';
@import 'flex';
@import 'group';
@import 'icon-btn';
@import 'input';
@import 'message';
@import 'modal';
@import 'post-form';
@import 'scroll-btn';

html,
body {
  overflow: hidden;
}

.loader {
  position: absolute;
  top: 45%;
  left: 45%;
}
.loader-overlay {
  position: absolute;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.7);
}
