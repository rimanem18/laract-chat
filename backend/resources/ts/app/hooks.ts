import React, { useCallback, useEffect, useMemo, useState } from 'react'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { fetchMessages, updateMessages } from '../slices/ChatMessagesSlice'
import type { RootState, AppDispatch } from './store'

import {
  chatMessageIdsSelector,
  chatMessagesEntitiesSelector,
  chatMessagesPromiseSelector,
} from '../selectors/ChatMessagesSelector'
import {
  userEmailSelector,
  userIdSelector,
  userNameSelector,
  userPromiseSelector,
} from '../selectors/UserSelector'
import {
  authEmailSelector,
  authMessageSelector,
  authNameSelector,
  authPasswordSelector,
  authPromiseSelector,
} from '../selectors/AuthSelector'
import {
  postContentSelector,
  postPromiseSelector,
  postUserIdSelector,
} from '../selectors/PostSelector'
import {
  groupIdsSelector,
  groupsEntitiesSelector,
  groupsOldestIdSelector,
  groupsPromiseSelector,
} from '../selectors/GroupsSelector'
import { fetchGroups } from '../slices/GroupsSlice'
import { useParams } from 'react-router'

// プレーンな useDispatch と useSelector の代わりにアプリ全体で使用する
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

// Auth Selector
export const useAuthName = () => {
  return useAppSelector(authNameSelector)
}
export const useAuthEmail = () => {
  return useAppSelector(authEmailSelector)
}
export const useAuthPassword = () => {
  return useAppSelector(authPasswordSelector)
}
export const useAuthMessage = () => {
  return useAppSelector(authMessageSelector)
}
export const useAuthPromise = () => {
  return useAppSelector(authPromiseSelector)
}

// User Selector
export const useUserId = () => {
  return useAppSelector(userIdSelector)
}
export const useUserEmail = () => {
  return useAppSelector(userEmailSelector)
}
export const useUserName = () => {
  return useAppSelector(userNameSelector)
}
export const useUserPromise = () => {
  return useAppSelector(userPromiseSelector)
}

// Groups Selector
export const useGroupsIds = () => {
  return useAppSelector(groupIdsSelector)
}
export const useGroupsEntities = () => {
  return useAppSelector(groupsEntitiesSelector)
}
export const useGroupsPromise = () => {
  return useAppSelector(groupsPromiseSelector)
}
export const useGroupsOldestId = () => {
  return useAppSelector(groupsOldestIdSelector)
}

// ChatMessages Selector
export const useChatMessageIds = () => {
  return useAppSelector(chatMessageIdsSelector)
}
export const useChatMessagesEntities = () => {
  return useAppSelector(chatMessagesEntitiesSelector)
}
export const useChatMessagesPromise = () => {
  return useAppSelector(chatMessagesPromiseSelector)
}

// Post Selector
export const usePostUserId = () => {
  return useAppSelector(postUserIdSelector)
}
export const usePostContent = () => {
  return useAppSelector(postContentSelector)
}
export const usePostPromise = () => {
  return useAppSelector(postPromiseSelector)
}

// Render
export const useScrollToBottom = (
  refObject: React.MutableRefObject<HTMLElement | null>
) => {
  const el = refObject.current
  if (el !== null) {
    el.scrollTo(0, el.scrollHeight)
  }
}

/**
 * 日付を 1900-01-01 00:00 形式にして返す
 * @param created_at
 * @returns
 */
export const useFormatDate = (created_at: string) => {
  // ゼロ埋め
  const zeroPadding = useCallback((num: number, len: number) => {
    return (Array(len).join('0') + num).slice(-len)
  }, [])

  // 日付フォーマット
  const date = useMemo(() => new Date(created_at), [])
  const year = date.getFullYear()
  const month = zeroPadding(date.getMonth() + 1, 2)
  const day = zeroPadding(date.getDate(), 2)
  const hour = zeroPadding(date.getHours(), 2)
  const min = zeroPadding(date.getMinutes(), 2)
  const datetime = `${year}/${month}/${day} ${hour}:${min}`

  return datetime
}

/**
 * メッセージが何もフェッチされていないときだけfetchする
 */
export const useInitFetchMessages = () => {
  const chatMessagesIds = useChatMessageIds()
  const chatMessagesPromise = useChatMessagesPromise()
  const dispatch = useAppDispatch()

  useEffect(() => {
    if (chatMessagesIds.length === 0 || chatMessagesPromise !== 'loading') {
      dispatch(fetchMessages())
    }
  }, [])
}

/**
 * メッセージ一覧の差分を取得して更新する
 */
export const useUpdateMessages = () => {
  const chatMessagesIds = useChatMessageIds()
  const chatMessagesPromise = useChatMessagesPromise()
  const postPromise = usePostPromise()
  const dispatch = useAppDispatch()

  useEffect(() => {
    if ([chatMessagesPromise, postPromise].every((v) => v === 'idle')) {
      dispatch(updateMessages())
    }
  }, [postPromise, chatMessagesIds.length])
}

/**
 * グループ一覧を初回だけフェッチする
 */
export const useFetchGroups = () => {
  const groupsPromise = useGroupsPromise()
  const dispatch = useAppDispatch()

  useEffect(() => {
    if (groupsPromise !== 'loading') {
      dispatch(fetchGroups())
    }
  }, [])
}

/**
 * URL から groupId を取得して返す
 * @returns string?
 */
export const useParamGroupId = () => {
  const { groupId } = useParams<{ groupId?: string }>()

  return groupId
}

export const useModalStyle = () => {
  const modalStyle = {
    overlay: {
      top: 0,
      left: 0,
      backgroundColor: 'rgba(0,0,0,0.85)',
    },
    content: {
      Position: 'relative',
      backgroundColor: '#f2f2f2',
      borderRadius: '1rem',
      padding: '0px',
      width: '30em',
      height: '15em',
      top: '50%',
      left: '50%',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
    },
  }

  return useMemo(() => {
    return modalStyle
  }, [])
}

export const useGroupModal = (groupName: string) => {
  const [isConfirm, setIsConfirm] = useState(false)
  const [isOpen, setIsOpen] = useState(false)
  const [newGroupName, setNewGroupName] = useState('')

  const openModal = useCallback(() => {
    setIsOpen(true)
  }, [])
  const closeModal = useCallback(() => {
    setIsOpen(false)
    setIsConfirm(false)
  }, [])
  const openConfirm = useCallback(() => {
    setIsConfirm(true)
  }, [])
  const closeConfirm = useCallback(() => {
    setIsConfirm(false)
  }, [])

  return [
    { isOpen, isConfirm, newGroupName },
    { openModal, closeModal, openConfirm, closeConfirm, setNewGroupName },
  ] as const
}

export const useDefaultGroupPath = () => {
  const oldestGroupsId = useGroupsOldestId().toString()
  const groupPromise = useGroupsPromise()
  const [path, setPath] = useState('')

  useEffect(() => {
    if (groupPromise !== 'loading') {
      setPath(`/groups/${oldestGroupsId}`)
    }
  }, [groupPromise])

  return path
}
