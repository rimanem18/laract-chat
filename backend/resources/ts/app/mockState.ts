import { RootState } from './store'
import { Group, Message, PromiseState, Role, RoleGroup } from './type'

type UseAuthState = {
  name: string
  email: string
  password: string
  message: string
  promise: PromiseState
}
type UseUserState = {
  id: number
  email: string
  name: string
  roles: string[]
  roleNumberIds: number[]
  promise: PromiseState
}
type UseChatMessagesState = {
  messages: {
    byId: Record<string, Message>
    allIds: string[]
  }
  promise: PromiseState
}
type UseGroupsState = {
  groups: {
    byId: Record<string, Group>
    allIds: string[]
    allNumberIds: number[]
  }
  roleGroup: {
    byId: Record<string, RoleGroup>
    allIds: string[]
  }
  promise: PromiseState
  oldestId: number
  defaultPath: string
}
type UsePostState = {
  userId: number
  content: string
  promise: PromiseState
}
type MockState = {
  authSlice: UseAuthState
  chatMessagesSlice: UseChatMessagesState
  groupsSlice: UseGroupsState
  userSlice: UseUserState
  menuSlice: { isOpen: false }
  postSlice: UsePostState
}

export const mockState: RootState = {
  authSlice: {
    name: '太郎',
    email: 'taro@example.com',
    password: 'authpassword',
    promise: 'idle',
    message: 'auth message',
  },
  chatMessagesSlice: {
    messages: {
      allIds: ['message1', 'message2'],
      byId: {
        message1: {
          id: 1,
          name: 'hoge',
          user_id: 1,
          group_id: 1,
          content: 'fugafuga',
          created_at: '1900-01-01',
          roles: ['role1'],
        },
        message2: {
          id: 2,
          name: 'piyo',
          user_id: 2,
          group_id: 2,
          content: 'fugafuga',
          created_at: '1900-01-01',
          roles: ['role1', 'role2'],
        },
      },
    },
    roleUser: {
      byId: {},
      allIds: [],
    },

    promise: 'idle',
  },
  groupsSlice: {
    groups: {
      byId: {
        group1: {
          id: 1,
          name: 'hello',
          roles: ['role1'],
        },
        group2: {
          id: 2,
          name: 'world',
          roles: ['role1', 'role2'],
        },
      },
      allIds: ['group1', 'group2'],
    },
    roleGroup: {
      byId: {},
      allIds: [],
    },
    promise: 'idle',
    oldestId: 1,
  },
  menuSlice: {
    isOpen: false,
  },
  postSlice: {
    userId: 1,
    content: 'fugafuga',
    promise: 'idle',
  },
  rolesSlice: {
    roles: {
      byId: {
        role1: {
          id: 1,
          name: 'staff',
          color: '#999999',
        },
        role2: {
          id: 2,
          name: 'admin',
          color: '#cccccc',
        },
      },
      allIds: ['role1', 'role2'],
    },
    promise: 'idle',
  },
  userSlice: {
    id: 1,
    name: '太郎',
    email: 'taro@example.com',
    roles: ['role1'],
    promise: 'idle',
  },
}
