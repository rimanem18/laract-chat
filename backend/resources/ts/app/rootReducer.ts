import { combineReducers } from '@reduxjs/toolkit'
import chatMessagesSlice from '../features/ChatMessagesSlice'
import userSlice, { authApi } from '../features/UserSlice'

const rootReducer = combineReducers({
  chatMessagesSlice,
  userSlice,
})

export type RootState = ReturnType<typeof rootReducer>
export default rootReducer
