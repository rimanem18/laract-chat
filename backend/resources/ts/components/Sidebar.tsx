import React from 'react'
import { Button, Grid } from '@mui/material'
import LogoutIcon from '@mui/icons-material/Logout'
import { useAppDispatch } from '../app/hooks'
import { logout } from '../slices/AuthSlice'
import Group from './Group'
import UserInfo from './UserInfo'
import { useHistory } from 'react-router'

const Sidebar = () => {
  const dispatch = useAppDispatch()

  const logoutHandler = () => {
    dispatch(logout())
  }

  return (
    <>
      <Grid container sx={{ my: 2 }}>
        <Grid container>
          <UserInfo />
        </Grid>
        <Grid container>
          <Button
            onClick={logoutHandler}
            startIcon={<LogoutIcon />}
            sx={{ ml: 1 }}
          >
            ログアウト
          </Button>
        </Grid>
      </Grid>
      <Group />
    </>
  )
}

export default React.memo(Sidebar)
