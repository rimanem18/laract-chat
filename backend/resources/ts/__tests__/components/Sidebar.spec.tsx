import React from 'react'
import '@testing-library/jest-dom'
import { fireEvent, render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'
import Sidebar from '../../components/Sidebar'
import { mockState } from '../../app/mockState'

const mockUseAppDispatch = jest.fn()
jest.mock('../../app/hooks', () => ({
  useAppDispatch:
    () =>
    (...args: any[]) =>
      mockUseAppDispatch(...args),
  useUserState: () => useUserStateMock(),
  useGroupsState: () => useGroupsStateMock(),
  useParamGroupId: () => useParamGroupIdMock(),
  useGroupModal: () => useGroupModalMock(),
}))

// Mock の定義
const user = mockState.userSlice
const groupState = mockState.groupsSlice

const useUserStateMock = jest.fn().mockReturnValue({
  userId: user.id,
  userName: user.name,
  userEmail: user.email,
  userRoleIds: user.role.ids,
  userRoleEntities: user.role.entities,
  userPromise: user.promise,
})
const useGroupsStateMock = jest.fn().mockReturnValue(groupState)
const useParamGroupIdMock = jest.fn().mockReturnValue(1)

let useGroupModalMock = jest.fn()

const Component = (
  <Router>
    <Sidebar />
  </Router>
)

const setup = () => {
  const screen = render(Component)
  const userName = screen.getByTestId('user-name')
  return {
    userName,
    ...screen,
  }
}

describe('Sidebar', () => {
  useGroupModalMock = jest.fn().mockReturnValue([
    {
      isOpen: false,
      isConfirm: false,
      newGroupName: groupState.groups.byId.group1.name,
    },
    {
      openModal: jest.fn(),
      closeModal: jest.fn(),
      openConfirm: jest.fn(),
      closeConfirm: jest.fn(),
      setNewGroupName: jest.fn(),
    },
  ])

  it('ユーザ名の一文字目が表示される', () => {
    const { userName } = setup()

    expect(userName.textContent).toBe('太')
  })

  it('snapshot', () => {
    const tree = renderer.create(Component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
