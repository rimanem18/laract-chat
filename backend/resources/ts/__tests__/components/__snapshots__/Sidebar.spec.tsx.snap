// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar snapshot 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-161s7uk-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-utr0cb-MuiAvatar-root"
        data-testid="user-name"
        title="太郎"
      >
        太
      </div>
      <div
        className="MuiBox-root css-7k1nrj"
      >
        <div
          className="MuiBox-root css-0"
        >
          太郎
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-w99pb7-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LogoutIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
        </span>
        ログアウト
      </button>
    </div>
  </div>,
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ControlPointIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.49 2 2 6.49 2 12s4.49 10 10 10 10-4.49 10-10S17.51 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
        />
      </svg>
    </span>
    新しいグループを追加
  </button>,
  <ul
    className="MuiList-root MuiList-padding css-x9esj-MuiList-root"
  >
    <div
      className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
      data-testid="group1"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          <div
            className="MuiBox-root css-24ohs5"
          >
            hello
          </div>
        </span>
      </div>
    </div>
    <div
      className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
      data-testid="group2"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          <div
            className="MuiBox-root css-24ohs5"
          >
            world
          </div>
        </span>
      </div>
    </div>
  </ul>,
]
`;
